src/llama_recipes/utils/train_utils.py

```diff
@@ -74,19 +74,29 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
         train_step_loss = []
         val_step_loss = []
         val_step_perplexity = []
-        
+
     epoch_times = []
     checkpoint_times = []
     results = {}
     best_val_loss = float("inf")
+    total_train_steps = 0
     for epoch in range(train_config.num_epochs):
```

Comment:
@wukaixingxp I wonder if we should remove some of the redundancy, maybe something like?

```
max_steps_reached = False  # Flag to indicate max training steps reached

for epoch in range(train_config.num_epochs):
    if max_steps_reached:
        break  # Stop starting a new epoch if max steps reached

    epoch_start_time = time.perf_counter()
    with MemoryTrace():  
        model.train()
        total_loss = 0.0
        total_length = len(train_dataloader) // gradient_accumulation_steps
        pbar = tqdm(colour="blue", desc=f"Training Epoch: {epoch+1}", total=total_length, dynamic_ncols=True)

        for step, batch in enumerate(train_dataloader):
            total_train_steps += 1

            # Check if max training steps reached
            if train_config.max_train_step > 0 and total_train_steps > train_config.max_train_step:
                print(f"Max training steps reached at step {total_train_steps}. Stopping training.")
                max_steps_reached = True
                break  # Exit batch loop

if max_steps_reached and not train_config.enable_fsdp or local_rank == 0:
    print("Training stopped after reaching the maximum number of steps.")
  ```

Reply:
I changed my code accordingly, thanks for the feedback.